src/main/host/syscall/unistd.rs

```diff
@@ -93,214 +91,172 @@ pub fn dup_helper(
         // if it's a legacy descriptor, use the C syscall handler instead
         CompatDescriptor::Legacy(_) => unsafe {
             c::syscallhandler_write(sys as *mut c::SysCallHandler, args as *const c::SysCallArgs)
+                .into()
         },
     }
 }
 
-pub fn pwrite64(sys: &mut c::SysCallHandler, args: &c::SysCallArgs) -> c::SysCallReturn {
-    let fd = unsafe { args.args[0].as_i64 } as libc::c_int;
-    let buf_ptr = unsafe { args.args[1].as_ptr };
-    let buf_size = unsafe { args.args[2].as_u64 } as libc::size_t;
-    let offset = unsafe { args.args[3].as_i64 } as libc::off_t;
+pub fn pwrite64(sys: &mut c::SysCallHandler, args: &SysCallArgs) -> SyscallReturn {
+    let fd = libc::c_int::from(args.get(0));
+    let buf_ptr = PluginPtr::from(args.get(1));
+    let buf_size = libc::size_t::from(args.get(2));
+    let offset = libc::off_t::from(args.get(3));
 
     // get the descriptor, or return early if it doesn't exist
-    let desc = match syscall::get_descriptor(fd, sys.process) {
-        Ok(d) => unsafe { &mut *d },
-        Err(errno) => return SyscallReturn::Error(errno).into(),
-    };
+    let desc = unsafe { &mut *syscall::get_descriptor(fd, sys.process)? };
 
     match desc {
         CompatDescriptor::New(desc) => write_helper(sys, fd, desc, buf_ptr, buf_size, offset),
         // if it's a legacy descriptor, use the C syscall handler instead
         CompatDescriptor::Legacy(_) => unsafe {
-            c::syscallhandler_pwrite64(sys as *mut c::SysCallHandler, args as *const c::SysCallArgs)
+            c::syscallhandler_pwrite64(sys as *mut c::SysCallHandler, args as *const SysCallArgs)
+                .into()
         },
     }
 }
 
 fn write_helper(
-    sys: &mut c::SysCallHandler,
+    _sys: &mut c::SysCallHandler,
     _fd: libc::c_int,
     desc: &Descriptor,
-    buf_ptr: c::PluginPtr,
+    buf_ptr: PluginPtr,
     buf_size: libc::size_t,
     offset: libc::off_t,
-) -> c::SysCallReturn {
+) -> SyscallReturn {
     // TODO: dynamically compute size based on how much data is actually available in the descriptor
     let size_needed = std::cmp::min(buf_size, c::SYSCALL_IO_BUFSIZE as usize);
 
-    let buf = match syscall::get_readable_ptr::<u8>(sys.process, sys.thread, buf_ptr, size_needed) {
-        Ok(ptr) => unsafe { Some(std::slice::from_raw_parts(ptr, size_needed)) },
-        Err(PluginPtrError::Null) | Err(PluginPtrError::ZeroLen) => {
-            match (buf_ptr.val == 0, buf_size) {
-                // non-null and zero len
-                (false, 0) => Some(&[][..]),
-                // null and zero len
-                (true, 0) => return SyscallReturn::Success(0).into(),
-                // null and non-zero len
-                (true, _) => None,
-                // non-null and non-zero len
-                (false, _) => unreachable!(),
-            }
-        }
-        Err(PluginPtrError::LenTooSmall) => unreachable!(),
-    };
```

Comment:
I don't think we want to remove this logic for a few reasons:

1. If we ignore it, we can end up with undefined behaviour like mentioned in a previous comment.
2. Our test cases for pipes are very limited and I don't think any test scenarios with NULL buffers and zero-length buffers. Without this logic I don't think they are working correctly. For example with some descriptor types such as sockets (and maybe pipes since we have no tests for it) there's no error if you call `read()` with a NULL buffer and non-zero length, and if the descriptor has no data to read. But the current code will always return an `EFAULT` before `posix_file.borrow_mut().read()` is called.